### PR TITLE
[quantization] Support qgemm int8 for BladeDISC on GPU

### PR DESCRIPTION
**Pull request by linbinskn**
to #567 
- [x] const weight transpose pass
- [x] bladnn int8 gemm library call
- [x] end2end subgraph test (only work when blade_gemm is on)
- [x] qconv related bug fix

### CODE REVIEW

File: tao_compiler/mlir/xla/ral/context/stream_executor_based_impl.cc [hunk 1]
```diff
@@ -401,6 +404,64 @@ void ral_gemm(ExecutionContext* ctx, void* stream_handle, MemRefType<InT, 2> A,
...
+  TAO_VLOG(0) << "weight.sizes[0]: " << weight.sizes[0]
+              << " weight.sizes[1]: " << weight.sizes[1];
+  TAO_VLOG(0) << "result.sizes[0]: " << result.sizes[0]
+              << " result.sizes[1]: " << result.sizes[1];
+
+  auto gpu_driver = ctx->getDriver<GPUDriver>(GPUDriver::name());
+  float input_scale, weight_scale, result_scale;
+  gpu_driver->d2h(ctx, stream_handle, &inputScales.data[0], &input_scale,
+                  sizeof(float));
+  gpu_driver->d2h(ctx, stream_handle, &weightScales.data[0], &weight_scale,
```
Comment (wyzero) [inline tao_compiler/mlir/xla/ral/context/stream_executor_based_impl.cc:431]: I see the UT using per-tensor quantization  for weight while we only handle the first element of the scale. Is it the supposed behavior?

File: tao_compiler/mlir/disc/transforms/quantized_dot_rewriter.cc [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+
+  LogicalResult matchAndRewrite(mlir::mhlo_disc::QuantizedDotGeneralOp op,
+                                PatternRewriter& rewriter) const override {
+    bool onGpu = placement_utils::isGpuMhlo(op);
+    // Only need to transpose weight on gpu currently
+    if (onGpu == false) {
+      return failure();
+    }
+
+    if (op.weight().getDefiningOp<mhlo::TransposeOp>()) {
```
Comment (wyzero) [inline tao_compiler/mlir/disc/transforms/quantized_dot_rewriter.cc:97]: The candidate `QuantizedDotGeneralOp` ops are those having `k*n` layout, not those having `TransposeOp` operands. It's not reliable to only check the existence of the transpose op (e.g. transpose + dot op having n x k layout should also be taken into consideration).

File: tao_compiler/mlir/xla/ral/context/stream_executor_based_impl.cc (hunk 1 above)
Comment (linbinskn) [reply tao_compiler/mlir/xla/ral/context/stream_executor_based_impl.cc:431]: Yes, we only support per tensor quantization on gpu. This ut is used for per tensor testing although its weight scale is per channel.

File: tao_compiler/mlir/disc/transforms/quantized_dot_rewriter.cc [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+      return failure();
+    }
+
+    auto weightTy = op.weight().getType().dyn_cast<RankedTensorType>();
+    int rank = weightTy.getRank();
+
+    // This transpose pass can only support dot with rank equal to 2
+    if (rank != 2) {
```
[the remark this answers is not in the document]
Comment (linbinskn) [reply tao_compiler/mlir/disc/transforms/quantized_dot_rewriter.cc]: Done

File: tao_compiler/mlir/disc/transforms/quantized_dot_rewriter.cc (hunk 2 above)
Comment (linbinskn) [reply tao_compiler/mlir/disc/transforms/quantized_dot_rewriter.cc]: Make sense, done.

File: tao_compiler/mlir/xla/ral/context/stream_executor_based_impl.cc (hunk 1 above)
Comment (wyzero) [reply tao_compiler/mlir/xla/ral/context/stream_executor_based_impl.cc:431]: Then why not use a per-tensor UT directly？the compiler side is supposed to be ok for both cases. And the register kernel should use `memref<f32>` for weight scale/zero_point inputs.

Comment (linbinskn) [reply tao_compiler/mlir/xla/ral/context/stream_executor_based_impl.cc:431]: Of course we can use a per-tensor UT directly. However, I think per-channel/per-tensor should be unified in library call interface although we may launch different kernels for them. And if we support per-channel quantization later, we only need to modify ral kernel library call function.

**Review by wyzero — APPROVED** [review]

**Pull request merged**